gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1228 @@
+			}
+		}
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	state = funcctx->user_fctx;
+
+	if (state->entry_current_row < state->entry_processed)
+	{
+		HeapTuple	inputTuple = state->entry_result->vals[state->entry_current_row];
+		TupleDesc	inputTupleDesc = state->entry_result->tupdesc;
+
+		for (int i = 0; i < funcctx->tuple_desc->natts; i++)
+		{
+			values[i] = SPI_getbinval(inputTuple, inputTupleDesc, i + 1, &nulls[i]);
+		}
+		HeapTuple	resultTuple = heap_form_tuple(funcctx->tuple_desc, values, nulls);
+
+		state->entry_current_row++;
+		SRF_RETURN_NEXT(funcctx, HeapTupleGetDatum(resultTuple));
+	}
+
+	SPI_finish();
+
+	while (state->current_result < state->cdb_results.numResults)
+	{
+		struct pg_result *pgresult = state->cdb_results.pg_results[state->current_result];
+
+		if (pgresult)
+		{
+			int			nrows = PQntuples(pgresult);
+			int			ncols = PQnfields(pgresult);
+
+			if (state->current_row < nrows)
+			{
+				for (int col = 0; col < ncols; col++)
+				{
+					if (PQgetisnull(pgresult, state->current_row, col))
+					{
+						values[col] = (Datum)0;
+						nulls[col] = true;
+					}
+					else
+					{
+						char	   *value = PQgetvalue(pgresult, state->current_row, col);
+
+						values[col] = InputFunctionCall(&state->inputFuncInfos[col], value, state->typIOParams[col], -1);
+					}
+				}
+				result = heap_form_tuple(funcctx->tuple_desc, values, nulls);
+				state->current_row++;
+				SRF_RETURN_NEXT(funcctx, HeapTupleGetDatum(result));
+			}
+			else
+			{
+				state->current_row = 0;
+				state->current_result++;
+			}
+		}
+		else
+			state->current_result++;
+	}
+
+	SRF_RETURN_DONE(funcctx);
+}
+
+static void
+track_db(Oid dbid, bool reg)
+{
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		AlterDatabaseSetStmt stmt;
+		VariableSetStmt v_stmt;
+		A_Const		aconst =
+		{.type = T_A_Const,.val = {.type = T_String,.val.str = reg ? "t" : "f"}};
+
+		stmt.type = T_AlterDatabaseSetStmt;
+		stmt.dbname = get_database_name(dbid);
+
+		if (stmt.dbname == NULL)
+			elog(ERROR, "[arenadata_toolkit] database %u does not exist", dbid);
+
+		stmt.setstmt = &v_stmt;
+
+		v_stmt.type = T_VariableSetStmt;
+		v_stmt.kind = VAR_SET_VALUE;
+		v_stmt.name = "arenadata_toolkit.tracking_is_db_tracked";
+		v_stmt.args = lappend(NIL, &aconst);
+		v_stmt.is_local = false;
+
+		tf_guc_unlock_tracked_once();
+
+		AlterDatabaseSet(&stmt);
+	}
+
+	if (!reg)
+		bloom_set_unbind(&tf_shared_state->bloom_set, dbid);
+	else if (!bloom_set_bind(&tf_shared_state->bloom_set, dbid))
+		elog(ERROR, "[arenadata_toolkit] exceeded maximum number of tracked databases");
+}
+
+/*
+ * Registers current (if dbid is 0) or specific database as tracked by arenadata_toolkit tables tracking.
+ * Dispatches call to segments by itself. Binds a bloom filter to the registered database if possible.
+ */
+Datum
+tracking_register_db(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = PG_GETARG_OID(0);
+
+	dbid = dbid == InvalidOid ? MyDatabaseId : dbid;
```

Comment:
```suggestion
	dbid = (dbid == InvalidOid) ? MyDatabaseId : dbid;
```

And in similar places as well.

Reply:
Fixed